File: ssp/spnlms.c
/* SPNLMS     11/13/85 */
/* IMPLEMENTS NLMS ALGORITHM B(K+1)=B(K)+2*MU*E*X(K)/((L+1)*SIG) */
/* X(O:N-1)=DATA VECTOR      INPUT SENT     OUTPUT RETURNED */
/* D(0:N-1)=DESIRED SIGNAL VECTOR */
/* N SPECIFIES NUMBER OF DATA POINTS IN X AND D */
/* B(0:L)=ADAPTIVE COEFFICIENTS OF LTH ORDER FIR FILTER */
/* MU=CONVERGENCE PARAMETER - DECLARE REAL */
/* SIG=INPUT SIGNAL POWER ESTIMATE - UPDATED INTERNALLY */
/* AL=FORGETTING FACTOR   SIG(K)=AL*(X(K)**2)+(1-AL)*SIG(K-1) */
/* IERROR=0      NO ERRORS DETECTED */
/*        1      INVALID ORDER   L<0 */
/*        2      INVALID CONVERGENCE PARAMETER   MU<=0 OR >=1 */
/*        3      INPUT POWER ESTIMATE  SIG<=0 */
/*        4      FORGETTING FACTOR   AL<0 OR =>1 */
/*        5      RESPONSE EXCEEDS 1.E10 */

#ifndef KR
void spnlms(float *x, long *n, float *d, float *b, long *l, float *mu, float *sig, float *al, float *px, long *error)
#else
void spnlms(x, n, d, b, l, mu, sig, al, px, error)
long *n, *l, *error;
float *x, *d, *b, *mu, *sig, *al, *px;
#endif
{
    /* Local variables */
    long k, ll;
    float e, tmp;

    if (*l < 0)
    {
	*error = 1;
	return;
    }
    if (*mu <= 0.0 || *mu >= 1.0)
    {
	*error = 2;
	return;
    }
    if (*sig <= 0.0)
    {
	*error = 3;
	return;
    }
    if (*al < 0.0 || *al >= 1.0)
    {
	*error = 4;
	return;
    }

    for (k = 0 ; k < *n ; ++k)
    {
	px[0] = x[k];
	x[k] = 0.0;

	for (ll = 0 ; ll <= *l ; ++ll)
	{
	    x[k] += b[ll] * px[ll];
	}

	if (ABS(x[k]) > BIG)
	{
	    *error = 5;
	    return;
	}

	e = d[k] - x[k];
	*sig = *al * (px[0] * px[0]) + (1.0 - *al) * *sig;
	tmp = *mu * 2 / ((1.0 + (float) *l) * *sig);

	for (ll = 0 ; ll <= *l ; ++ll)
	{
	    b[ll] += tmp * e * px[ll];
	}

	for (ll = *l ; ll >= 1 ; --ll)
	{
	    px[ll] = px[ll - 1];
	}
    }

    *error = 0;
    return;
} /* spnlms */

File: ssp/spnlms.m
function [y,b,px]=spnlms(x,d,b,mu,sigma,alpha,px)
% [y,b,px]=spnlms(x,d,b,mu,sigma,alpha,px)
% NLMS algorithm. See Fig. 12.2. x,d=input and desired row
% vectors; b=LMS weight row vector; mu,sigma,alpha as in
% (12.8); px=row vector with past x values; may be "px=0".
% Inputs:
%    x      = input data row vector [x(1),x(2),...,x(N)].
%    d      = desired signal row vector, same length as x.
%    b      = row vector with L+1 adaptive FIR weights. 
%             H(z)=b(1)+b(2)*z^(-1)+...+b(L+1)*z^(-L).
%    mu     = convergence parameter in eq. 12.8; scalar.
%    sigma  = estimate of x^2 (updated); scalar.
%    alpha  = forgetting factor in eq. 12.8; scalar.
%    px     = past values of x row vector (may be "0").
% Outputs:  
%    y  = output row data vector with epsilon=d-y.
%    b  = updated adaptive weight row vector, length L+1.
%    px = updated px row vector; [x(N-1),...,x(N-L)].

N=length(x); L=length(b)-1;
if N~=length(d),
   error('SPNLMS: lengths of x and d row vectors not equal.');
end
if (mu<=0)|(mu>=1)|(sigma<=0)|(alpha<0)|(alpha>=1),
   error('SPNLMS: mu, sigma, or alpha out of range.');
end
y=zeros(1,N);
if(length(px)<L),
   px=[px,zeros(1,L-length(px))];
end
px=[0,px];
for k=1:N,
   px(1)=x(k);
   y(k)=b*px';
   if abs(y(k))>1e10,
      fprintf('\nSPNLMS warning: |y| output > 1e10.\n');
      y(k+1:N)=zeros(1,N-k);
      return
   end
   e=d(k)-y(k);
   sigma=alpha*(px(1)^2)+(1-alpha)*sigma;
   tmp=2*mu/((L+1)*sigma);
   b=b+tmp*e*px;
   px(L+1:-1:2)=px(L:-1:1);
end
px=px(2:L+1);
return
